Iterable validation and flatmap

EivindAntonsen opened the pull request:
Two suggested extension functions.

LukasForst commented on src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:
```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
+    return flatMapIndexedTo(ArrayList(), function)
+}
+
+private inline fun <T, R, C : MutableCollection<in R>> Iterable<T>.flatMapIndexedTo(destination: C,
```
I'd suggest making this function public as the other `mapTo` functions are public - thus allowing inlining of `flatMapIndexedNotNull`

LukasForst replied:
keep in mind, that all public functions must be documented

EivindAntonsen replied:
Added documentation similar to the standard kotlin lib as the functions themselves are really pretty similar.

LukasForst commented on src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:
```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
+
+private inline fun <T, R, C : MutableCollection<in R>> Iterable<T>.flatMapIndexedTo(destination: C,
+                                                                                    transform: (index: Int, T) -> Iterable<R>?): C {
+
```
please format the file properly using IntelliJ (I intent to add some formatting/linting tool like detekt, but I haven't had time yet)
```
inline fun <T, R, C : MutableCollection<in R>> Iterable<T>.flatMapIndexedTo(
    destination: C,
    transform: (index: Int, T) -> Iterable<R>?
): C {
    forEachIndexed { index, element ->
        transform(index, element)?.let { elements ->
            destination.addAll(elements)
        }
    }
    
    return destination
}
```

EivindAntonsen replied:
forgot to add code style schema for this fork, will correct it

LukasForst replied:
> understandings of properly

heh, true, `properly` is not correct wording 😄  -- I'm using the formatting, IntelliJ is using, so I just let IntelliJ format everything before I commit it

EivindAntonsen replied:
I hit format on both of those files, and it seems to have resulted in a lot of changes.

LukasForst replied:
this formatting looks good to me 👍

LukasForst commented on src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:
```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
+ *
+ * @param validationFunction is the accumulator function that verifies the elements.
+ */
+fun <S, T : S> Iterable<T>.validateElements(validationFunction: (acc: S, T) -> Boolean): Boolean {
```
I'm not sure about name `validateElements` when this function provide possibility to validate relationship between ordered elements - so you're in a fact verifying that `n` and possibly `n+1` have some relationship

to me `validateElements` sounds like it provides a way how to validate collection of elements without any additional relationships

EivindAntonsen replied:
Agreed. Will find something more fitting.

EivindAntonsen replied:
I think `foldValidated` is the closest match I can think of. It's essentially a fold with early return, no?

LukasForst replied:
sounds gut to me

LukasForst commented on src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt:
```diff
@@ -224,4 +224,24 @@ internal class IterableExtensionsTest {
+    }
+
+    @Test
+    fun flatMapIndexedNotNull() {
```
could you please add a test which utilizes filtering `null`? thanks!

EivindAntonsen replied:
Hi, in the recent commit I did include something that checked for nulls in the end result, but I'm not sure that is what you meant. If it wasn't, can you elaborate?

LukasForst replied:
yup, that one, thanks!

LukasForst commented on src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt:
```diff
@@ -224,4 +224,24 @@ internal class IterableExtensionsTest {
+            nextNumber - previousNumber == 1
+        }
+
+        assertEquals(true, differenceIsMaxOne)
```
could you please add a test where the result is `false`? thanks!

LukasForst commented on src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:
```diff
@@ -497,3 +509,57 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
+ * function being invoked on each element of original collection.
+ */
+fun <T, R> Iterable<T>.flatMapIndexedNotNull(transform: (index: Int, T) -> Iterable<R>?): List<R> {
+    return flatMapIndexedTo(ArrayList(), transform)
```
as the `flatMapIndexedTo` is inlined, this one can be too, please inline it, thanks!

LukasForst commented on src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:
```diff
@@ -497,3 +509,57 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
+ *
+ * @param validationFunction is the accumulator function that verifies the elements.
+ */
+fun <S, T : S> Iterable<T>.foldValidated(validationFunction: (acc: S, T) -> Boolean): Boolean {
```
`inline` would be nice as well

LukasForst reviewed (changes requested):
Hey, thanks for the update! Resolved everything, two more comments, about adding `inline` keyword.

Also, could you please rebase it on top of current master? There are some conflicts, thank you! Once conflicts are resolved, I'll approve the code.

EivindAntonsen commented:
Hey I rebased before I added the inline (never done a rebase before, we've only used merge at my job), so there's one extra commit after. Does this mean I need another rebase?

LukasForst commented:
> Hey I rebased before I added the inline (never done a rebase before, we've only used merge at my job), so there's one extra commit after. Does this mean I need another rebase?
thanks! now is just necessary to resolve conflicts with the master branch

[rebase](https://medium.com/datadriveninvestor/git-rebase-vs-merge-cc5199edd77c) is nice thing to do as it preserves linear history, however I don't enforce it, if you'd like to resolve the commits with merge, feel free

Right now, there's conflict in the `IterableExtensions` with current master branch -- there's an example how to resolve conflicts between forks [here](https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks/38955036).

EivindAntonsen commented:
Hm, my first time working on a fork here, not sure I did this right.

I merged this branch into master on my forked project, and then pushed. But it still does not look right here. Does that mean I should've made a pr from my forked master instead?

LukasForst commented:
It's not necessary to make PR from your master branch, but it is necessary to rebase/merge the master branch of this repo to your `iterable-validation-and...` branch. 

Either by using [this](https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks/38955036) approach or by updating your master directly using github (or at least I think it should be possible to update your master from my master using github)

LukasForst commented:
You can create PR from my master to yours using this https://github.com/EivindAntonsen/katlib/compare/master...LukasForst:master link, once you have updated master, you merge it to your branch, which updates this PR and will be ready to be merged

EivindAntonsen commented:
I think this should be it. :)
Let me know if something needs changing.

LukasForst commented:
Oh, I looked into the logs and the Github action could not connect to gradle plugin repository, they have some problems, all green now! Great.

LukasForst reviewed (approved):
Approved, thank you!
